k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,249 @@
+ *  Device cannot let host terminate/exit the process. And calling a
+ *  `__host__ function("exit")` from a `__host__ __device__ function`
+ *  is not allowed. Thus, the exit(e) is put here other than inside of
+ *  `K2_CUDA_DEBUG`.
+ *
+ * @todo
+ *  Should the error code return by this macro?
+ */
+#ifndef K2_CUDA_CHECK_ERROR
+  #define K2_CUDA_CHECK_ERROR(e)                                   \
+    if (k2::K2_CUDA_DEBUG((cudaError_t)(e), __FILE__, __LINE__)) { \
+      exit(e);                                                     \
+    }
+#endif
+
+/**
+ * @def K2_CUDA_API_SAFE_CALL(CUDA_RUNTIME_API)
+ *
+ * @brief Macro for checking cuda standard runtime api return status.
+ *
+ * @details
+ *  If api return status is not cudaSuccess, print and exit.
+ *  Thus it's actually could be an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * @note
+ *  All runtime api return an error code. But for an asynchronous api,
+ *  the error code only reports errors that occur on the host
+ *  prior to executing the task, typically related to parameter validation.
+ *  The device kernel error would unfortunately be left for next runtime api
+ *  call. To check the whole process of cuda runtime async api,
+ *  as cuda kernel launches are asynchronous also,
+ *  the same macro for kernel could be used. See
+ *  `K2_CUDA_KERNEL_SAFE_CALL`.
+ *
+ * @code{.cpp}
+ * K2_CUDA_API_SAFE_CALL(CUDA_RUNTIME_API())
+ * @endcode
+ */
+#ifndef K2_CUDA_API_SAFE_CALL
+  #define K2_CUDA_API_SAFE_CALL(c) \
+    do {                           \
+      cudaError_t e = (c);         \
+      K2_CUDA_CHECK_ERROR(e);      \
+    } while (0)
+#endif
+
+/**
+ * @def K2_CUDA_KERNEL_SAFE_CALL([kernel|cuda_runtime_api])
+ *
+ * @brief Macro for checking cuda standard runtime api return status.
+ *
+ * @details
+ *  If api return status is not cudaSuccess, print and exit.
+ *  Thus it's actually an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * @note
+ *  Kernel launches do not return any error code, thus checking should after it.
+ *  To wait kernel to finish, cudaDeviceSynchronize is called between.
+ *  And to avoid pre-launch error disturbing, check error before kernel launch
+ *  is also needed.
+ *  (But not necessary if programmer promise each error is take cared properly.)
+ *
+ * @remark
+ *  macro `__VA_ARGS__` is used to pass the kernel<<<...>>> as one argument,
+ *  otherwise the compiler rise a error "passed 2 arguments, but takes just 1".
+ *
+ * @code{.cpp}
+ * K2_CUDA_KERNEL_SAFE_CALL(kernel_func<<<...>>>())
+ * @endcode
+ */
+#ifndef K2_CUDA_KERNEL_SAFE_CALL
+  #define K2_CUDA_KERNEL_SAFE_CALL(...)               \
+    do {                                              \
+      K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+      (__VA_ARGS__);                                  \
+      K2_CUDA_API_SAFE_CALL(cudaDeviceSynchronize()); \
+      K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+    } while (0)
+#endif
+
+/**
+ * @def K2_DLOG
+ *
+ * @brief Log macro for printf statements.
```

Comment:
Please clarify how it is supposed to be used with an example!

Reply:
Updated. Please see the gtest about it. Though I'm afraid it's not as good as expected, I need to know the exactly expected usages so I can make changes accordingly.

Reply:
@songmeixu please make sure there is an example in the documentation!  People won't know to look in the test file.

Reply:
Now, the usage example is put @code tag as doxygen suggests.